bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/KeyValueStorageRocksDB.java

```diff
@@ -285,41 +290,57 @@ private RocksDB initializeRocksDBWithBookieConf(String basePath, String subPath,
 
     @Override
     public void close() throws IOException {
-        db.close();
-        if (cache != null) {
-            cache.close();
-        }
-        if (options != null) {
-            options.close();
-        }
-        optionSync.close();
-        optionDontSync.close();
-        optionCache.close();
-        optionDontCache.close();
-        emptyBatch.close();
+        writeLock.lock();
+        try {
+            closed = true;
+            db.close();
+            if (cache != null) {
+                cache.close();
+            }
+            if (options != null) {
+                options.close();
+            }
+            optionSync.close();
+            optionDontSync.close();
+            optionCache.close();
+            optionDontCache.close();
+            emptyBatch.close();
+        } finally {
+            writeLock.unlock();
+        }
     }
 
     @Override
     public void put(byte[] key, byte[] value) throws IOException {
+        readLock.lock();
```

Comment:
Acquiring the read lock for each put/get operation might be quite expensive, since the lock has to maintain the state of threads and the order they try to acquire the lock. 

We could use a different approach, like a reference counter on the entire object. When everyone is done using it, then rocksdb is really closed

Reply:
@AnonHxy It seems that a StampedLock could be sufficient here since there doesn't seem to be a need for re-entrancy of the lock. @merlimat would using a StampedLock address your concern?

Reply:
There's an alternative PR #4581 which simply handles `close` and `count`